apps/web/components/ui/tambo/chatwithtambo/tools/api-key-tools.ts

```diff
@@ -0,0 +1,129 @@
+      partiallyHiddenKey: z.string().nullable(),
+      lastUsedAt: z.date().nullable(),
+      projectId: z.string(),
+      hashedKey: z.string(),
+      createdAt: z.date(),
+      updatedAt: z.date(),
+      createdByUserId: z.string(),
+    }),
+  );
+
+/**
+ * Zod schema for the `deleteProjectApiKey` function.
+ * Defines arguments as the project ID string and API key ID string,
+ * and the return type as an object indicating the key was processed for deletion (e.g., `{ deletedKey: undefined }`).
+ */
+export const deleteProjectApiKeySchema = z
+  .function()
+  .args(
+    z.string().describe("The project ID"),
+    z.string().describe("The API key ID"),
+  )
+  .returns(
+    z.object({
+      deletedKey: z.void(),
+    }),
+  );
+
+/**
+ * Register API key management tools
+ */
+export function registerApiKeyTools(
+  registerTool: RegisterToolFn,
+  ctx: ToolContext,
+) {
+  /**
+   * Registers a tool to fetch all API keys for a specific project.
+   * @param {string} projectId - The project ID to fetch API keys for
+   * @returns {Object} Object containing an array of API key details
+   */
+  registerTool({
+    name: "fetchProjectApiKeys",
+    description: "get all api keys for the current project",
+    tool: async (projectId: string) => {
+      return await ctx.trpcClient.project.getApiKeys.query(projectId);
+    },
+    toolSchema: fetchProjectApiKeysSchema,
+  });
+
+  /**
+   * Registers a tool to generate a new API key for a project.
+   * @param {string} projectId - The project ID to generate an API key for
+   * @param {string} name - The name/label for the new API key
+   * @returns {Object} Generated API key details including the key value and metadata
+   */
+  registerTool({
+    name: "generateProjectApiKey",
+    description: "generate a new api key for the current project",
+    tool: async (projectId: string, name: string) => {
+      const result = await ctx.trpcClient.project.generateApiKey.mutate({
+        projectId,
+        name,
+      });
+
+      // Invalidate the API keys cache to refresh the component
+      await invalidateApiKeysCache(ctx, projectId);
+
+      return result;
+    },
+    toolSchema: generateProjectApiKeySchema,
+  });
+
+  /**
+   * Registers a tool to delete an existing API key from a project.
+   * @param {string} projectId - The project ID containing the API key
+   * @param {string} apiKeyId - The ID of the API key to delete
+   * @returns {Object} Confirmation message that the key was deleted
+   */
+  registerTool({
+    name: "deleteProjectApiKey",
+    description: "delete an api key for the current project",
+    tool: async (projectId: string, apiKeyId: string) => {
+      await ctx.trpcClient.project.removeApiKey.mutate({
+        projectId,
+        apiKeyId,
+      });
+
+      // Invalidate the API keys cache to refresh the component
+      await invalidateApiKeysCache(ctx, projectId);
+
+      return { deletedKey: "The key was deleted successfully." };
```

Comment:
curious why you can't just return the value from `.removeApiKey.mutate()` here?

Basically I think all of these actual tool calls should be very formulaic without special cases:


```ts
registerTool({
    name: hook.name,
    tool: async (...args) => ctx.trpcClient.foo.bar.mutate(...args);
   ...
```

That said, handling cache invalidations is tricky. maybe there's a way to define the invalidations along side the mutations somehow? hmmm will need to think on that